File: src/symbol_table.rs
use std::collections::HashMap;
use crate::ast::TuneStmt;
use crate::lexer::Token;

pub struct SymbolTable {
    pub tunes: HashMap<String, TuneStmt>,
    pub variables: HashMap<String, u32>,
}

impl SymbolTable {
    // Core Tune Functions

    pub fn add_tune(&mut self, identifier: Token, tune_stmt: TuneStmt) -> bool {
        if self.tunes.contains_key(identifier.literal.as_str()) {
            return false;
        }

        self.tunes.insert(identifier.literal, tune_stmt);
        return true;
    }

    pub fn get_tune(&mut self, identifier: Token) -> Option<&TuneStmt> {
        self.tunes.get(identifier.literal.as_str())
    }

    pub fn drop_tune(&mut self, identifier: Token) {
        self.tunes.remove(identifier.literal.as_str());
    }

    // Core Variable Functions

    pub fn add_variable(&mut self, identifier: Token, value: u32) -> bool {
        if self.variables.contains_key(identifier.literal.as_str()) {
            return false;
        }

        self.variables.insert(identifier.literal, value);
        return true;
    }

    pub fn get_variable(&mut self, identifier: Token) -> Option<&u32> {
        self.variables.get(identifier.literal.as_str())
    }

    pub fn reassign_variable(&mut self, identifier: Token, value: u32) -> bool {
        if !self.variables.contains_key(identifier.literal.as_str()) {
            return false;
        }

        self.variables.insert(identifier.literal, value);
        return true;
    }

    pub fn drop_variable(&mut self, identifier: Token) {
        self.variables.remove(identifier.literal.as_str());
    }

    // Utility Functions

    pub fn has_tune(&mut self, identifier: Token) -> bool {
        return if self.tunes.contains_key(identifier.literal.as_str()) {
            true
        } else {
            false
        }
    }

    pub fn has_variable(&mut self, identifier: Token) -> bool {
        return if self.variables.contains_key(identifier.literal.as_str()) {
            true
        } else {
            false
        }
    }
}

File: src/midilang/lang/SymbolTable.cpp
#include "SymbolTable.h"


MIDILang::SymbolTable::SymbolTable() {
    this->tunes = std::unordered_map<std::string, Tune>();
    this->variables = std::unordered_map<std::string, int>();
}

bool MIDILang::SymbolTable::addTune(MIDILang::Token identifier, MIDILang::Tune tuneStmt) {
    if (this->tunes.find(identifier.literal) != this->tunes.end()) return false;

    this->tunes.insert({ identifier.literal, tuneStmt });
    return true;
}

MIDILang::Tune MIDILang::SymbolTable::getTune(MIDILang::Token identifier) {
    return this->tunes.at(identifier.literal);
}

void MIDILang::SymbolTable::dropTune(MIDILang::Token identifier) {
    this->tunes.erase(identifier.literal);
}

bool MIDILang::SymbolTable::hasTune(MIDILang::Token identifier) {
    if (this->tunes.find(identifier.literal) != this->tunes.end()) return true;
    else return false;
}

bool MIDILang::SymbolTable::addVariable(MIDILang::Token identifier, int value) {
    if (this->variables.find(identifier.literal) != this->variables.end()) return false;

    this->variables.insert({ identifier.literal, value });
    return true;
}

int MIDILang::SymbolTable::getVariable(MIDILang::Token identifier) {
    return this->variables.at(identifier.literal);
}

bool MIDILang::SymbolTable::reassignVariable(MIDILang::Token identifier, int value) {
    if (this->variables.find(identifier.literal) == this->variables.end()) return false;

    this->variables.erase(identifier.literal);
    this->variables.insert({ identifier.literal, value });
    return true;
}

void MIDILang::SymbolTable::dropVariable(MIDILang::Token identifier) {
    this->variables.erase(identifier.literal);
}

bool MIDILang::SymbolTable::hasVariable(MIDILang::Token identifier) {
    if (this->variables.find(identifier.literal) != this->variables.end()) return true;
    else return false;
}
